File: kernel/user_programs/block_client.zig
const std = @import("std");
const fs_abi = @import("fs_abi.zig");
const block_protocol = @import("block_protocol.zig");
const service_registry_abi = @import("service_registry_abi.zig");

const syscall_alloc_page: u64 = 0x1;
const syscall_map_page: u64 = 0x2;
const syscall_send_cap: u64 = 0x6;
const syscall_grant_cap: u64 = 0x8;
const syscall_wait_event: u64 = 0x17;
const syscall_install_endpoint: u64 = 0x26;
const syscall_share_cap: u64 = 0x2B;
const syscall_signal_endpoint: u64 = 0x2C;

const page_right_cpu_read: u64 = 0x1;
const page_right_cpu_write: u64 = 0x2;
const default_response_poll_limit: u64 = 256;

pub const Error = error{
    RequestAllocFailed,
    RequestMapFailed,
    ResponseAllocFailed,
    ResponseMapFailed,
    ResponseGrantFailed,
    ConnectSendFailed,
    EndpointInstallFailed,
    Timeout,
    InvalidResponse,
    BufferTooSmall,
    Invalid,
    NotFound,
    NoRight,
    TooBig,
    IoError,
    Busy,
};

pub const ConnectOptions = struct {
    request_va: u64,
    response_va: u64,
    client_process_slot: u64,
    endpoint_id: u64,
    server_process_slot: u64,
    response_poll_limit: u64 = default_response_poll_limit,
};

pub const IdentifyResult = struct {
    block_size: u64,
    capacity_blocks: u64,
};

pub const Client = struct {
    request_va: u64,
    response_va: u64,
    request_paddr: u64,
    response_paddr: u64,
    server_endpoint_id: u64,
    root_token: u64,
    block_size: u64,
    capacity_blocks: u64,
    next_seq: u64 = 2,
    response_poll_limit: u64 = default_response_poll_limit,

    pub fn connect(options: ConnectOptions) Error!Client {
        const request_paddr = allocPage();
        if (request_paddr < 0x1000) return error.RequestAllocFailed;
        if (mapPage(options.request_va, request_paddr, true) != 0) return error.RequestMapFailed;

        const response_paddr = allocPage();
        if (response_paddr < 0x1000) return error.ResponseAllocFailed;
        if (mapPage(options.response_va, response_paddr, true) != 0) return error.ResponseMapFailed;
        if (grantCap(response_paddr, options.server_process_slot, page_right_cpu_read | page_right_cpu_write) != 0) {
            return error.ResponseGrantFailed;
        }

        var client = Client{
            .request_va = options.request_va,
            .response_va = options.response_va,
            .request_paddr = request_paddr,
            .response_paddr = response_paddr,
            .server_endpoint_id = options.endpoint_id,
            .root_token = 0,
            .block_size = 0,
            .capacity_blocks = 0,
            .response_poll_limit = options.response_poll_limit,
        };

        client.clearMappedPages();
        const request = client.requestHeader();
        request.magic = block_protocol.request_magic;
        request.version = block_protocol.version;
        request.op = block_protocol.opcodeRaw(.connect);
        request.object_token = 0;
        request.block_index = 0;
        request.block_count = 0;
        request.flags = 0;
        request.inline_bytes = 0;
        request.reserved0 = 0;
        request.reserved1 = 0;
        request.arg0 = response_paddr;
        request.arg1 = options.client_process_slot;
        compilerBarrier();
        request.request_seq = 1;

        if (shareCap(request_paddr, options.endpoint_id) != 0) return error.ConnectSendFailed;
        const response = try client.finishRequestOk(1, .connect);
        client.root_token = response.result_token;
        if (!fs_abi.isCapToken(client.root_token)) return error.InvalidResponse;
        client.block_size = response.arg0;
        client.capacity_blocks = response.arg1;
        if (client.block_size == 0 or client.capacity_blocks == 0) return error.InvalidResponse;
        return client;
    }

    pub fn connectFromRegistryPage(registry_page_va: u64, request_va: u64, response_va: u64, client_process_slot: u64) Error!Client {
        const entry = service_registry_abi.findService(registry_page_va, .block) orelse return error.NotFound;
        if (entry.process_slot == 0 or entry.endpoint_id == 0) return error.Invalid;
        if (installEndpoint(entry.endpoint_id, entry.process_slot) != 0) return error.EndpointInstallFailed;
        return connect(.{
            .request_va = request_va,
            .response_va = response_va,
            .client_process_slot = client_process_slot,
            .endpoint_id = entry.endpoint_id,
            .server_process_slot = entry.process_slot,
        });
    }

    pub fn connectFromServiceRegistry(request_va: u64, response_va: u64, client_process_slot: u64) Error!Client {
        return connectFromRegistryPage(service_registry_abi.page_va, request_va, response_va, client_process_slot);
    }

    pub fn identify(self: *Client) Error!IdentifyResult {
        const seq = try self.beginRequest(.identify, self.root_token, 0, 0, 0, &[_]u8{});
        const response = try self.finishRequestOk(seq, .identify);
        if (response.arg0 == 0 or response.arg1 == 0) return error.InvalidResponse;
        self.block_size = response.arg0;
        self.capacity_blocks = response.arg1;
        return .{
            .block_size = response.arg0,
            .capacity_blocks = response.arg1,
        };
    }

    pub fn readBlocks(self: *Client, block_index: u64, out: []u8) Error!usize {
        if (self.block_size == 0) return error.Invalid;
        if ((out.len % self.block_size) != 0) return error.BufferTooSmall;
        const block_count: usize = out.len / @as(usize, @intCast(self.block_size));
        if (block_count == 0 or block_count > std.math.maxInt(u32)) return error.BufferTooSmall;
        const seq = try self.beginRequest(.read_blocks, self.root_token, block_index, @intCast(block_count), 0, &[_]u8{});
        const response = try self.finishRequestOk(seq, .read_blocks);
        const inline_bytes: usize = response.inline_bytes;
        if (inline_bytes != out.len or inline_bytes > block_protocol.response_payload_bytes) return error.InvalidResponse;
        copyVolatileBytes(self.responsePayload(), out[0..inline_bytes]);
        return inline_bytes;
    }

    pub fn writeBlocks(self: *Client, block_index: u64, bytes: []const u8) Error!void {
        if (self.block_size == 0) return error.Invalid;
        if ((bytes.len % self.block_size) != 0) return error.BufferTooSmall;
        const block_count: usize = bytes.len / @as(usize, @intCast(self.block_size));
        if (bytes.len > block_protocol.request_payload_bytes or block_count == 0 or block_count > std.math.maxInt(u32)) {
            return error.BufferTooSmall;
        }
        const seq = try self.beginRequest(.write_blocks, self.root_token, block_index, @intCast(block_count), 0, bytes);
        _ = try self.finishRequestOk(seq, .write_blocks);
    }

    pub fn flush(self: *Client) Error!void {
        const seq = try self.beginRequest(.flush, self.root_token, 0, 0, 0, &[_]u8{});
        _ = try self.finishRequestOk(seq, .flush);
    }

    fn requestHeader(self: *const Client) *volatile block_protocol.BlockRequestHeader {
        return @ptrFromInt(self.request_va);
    }

    fn responseHeader(self: *const Client) *volatile block_protocol.BlockResponseHeader {
        return @ptrFromInt(self.response_va);
    }

    fn requestPayload(self: *const Client) [*]volatile u8 {
        return @ptrFromInt(self.request_va + block_protocol.request_header_bytes);
    }

    fn responsePayload(self: *const Client) [*]volatile u8 {
        return @ptrFromInt(self.response_va + block_protocol.response_header_bytes);
    }

    fn clearMappedPages(self: *const Client) void {
        clearPage(self.request_va);
        clearPage(self.response_va);
    }

    fn beginRequest(
        self: *Client,
        op: block_protocol.Opcode,
        object_token: u64,
        block_index: u64,
        block_count: u32,
        flags: u32,
        payload: []const u8,
    ) Error!u64 {
        if (payload.len > block_protocol.request_payload_bytes) return error.BufferTooSmall;
        self.clearMappedPages();
        const request = self.requestHeader();
        request.magic = block_protocol.request_magic;
        request.version = block_protocol.version;
        request.op = block_protocol.opcodeRaw(op);
        request.object_token = object_token;
        request.block_index = block_index;
        request.block_count = block_count;
        request.flags = flags;
        request.inline_bytes = @intCast(payload.len);
        request.reserved0 = 0;
        request.reserved1 = 0;
        request.arg0 = 0;
        request.arg1 = 0;
        copyBytesToVolatile(self.requestPayload(), payload);
        const seq = self.next_seq;
        self.next_seq += 1;
        compilerBarrier();
        request.request_seq = seq;
        _ = signalEndpoint(self.server_endpoint_id);
        return seq;
    }

    fn finishRequest(self: *Client, expected_seq: u64, expected_op: block_protocol.Opcode) Error!*volatile block_protocol.BlockResponseHeader {
        if (!self.waitForResponse(expected_seq)) return error.Timeout;
        const response = self.responseHeader();
        if (response.magic != block_protocol.response_magic or response.version != block_protocol.version) return error.InvalidResponse;
        if (response.op != block_protocol.opcodeRaw(expected_op) or response.response_seq != expected_seq) return error.InvalidResponse;
        _ = parseStatus(response.status) orelse return error.InvalidResponse;
        return response;
    }

    fn finishRequestOk(self: *Client, expected_seq: u64, expected_op: block_protocol.Opcode) Error!*volatile block_protocol.BlockResponseHeader {
        const response = try self.finishRequest(expected_seq, expected_op);
        const status = parseStatus(response.status) orelse return error.InvalidResponse;
        if (status != .ok) return statusToError(status);
        return response;
    }

    fn waitForResponse(self: *const Client, expected_seq: u64) bool {
        const response = self.responseHeader();
        var poll_count: u64 = 0;
        while (poll_count < self.response_poll_limit) : (poll_count += 1) {
            if (response.response_seq == expected_seq) return true;
            _ = waitEvent(false, 1);
        }
        return false;
    }
};

fn allocPage() u64 {
    return asm volatile (
        \\int $0x80
        : [ret] "={rax}" (-> u64),
        : [nr] "{rax}" (syscall_alloc_page),
        : .{ .rcx = true, .rdx = true, .r8 = true, .r9 = true, .r10 = true, .r11 = true, .memory = true });
}

fn mapPage(va: u64, paddr: u64, writable: bool) u64 {
    return asm volatile (
        \\int $0x80
        : [ret] "={rax}" (-> u64),
        : [nr] "{rax}" (syscall_map_page),
          [arg0] "{rdi}" (va),
          [arg1] "{rsi}" (paddr),
          [arg2] "{rdx}" (@as(u64, if (writable) 1 else 0)),
        : .{ .rcx = true, .rdx = true, .r8 = true, .r9 = true, .r10 = true, .r11 = true, .memory = true });
}

fn sendCap(paddr: u64, endpoint_id: u64) u64 {
    return asm volatile (
        \\int $0x80
        : [ret] "={rax}" (-> u64),
        : [nr] "{rax}" (syscall_send_cap),
          [arg0] "{rdi}" (paddr),
          [arg1] "{rsi}" (endpoint_id),
        : .{ .rcx = true, .rdx = true, .r8 = true, .r9 = true, .r10 = true, .r11 = true, .memory = true });
}

fn shareCap(paddr: u64, endpoint_id: u64) u64 {
    return asm volatile (
        \\int $0x80
        : [ret] "={rax}" (-> u64),
        : [nr] "{rax}" (syscall_share_cap),
          [arg0] "{rdi}" (paddr),
          [arg1] "{rsi}" (endpoint_id),
        : .{ .rcx = true, .rdx = true, .r8 = true, .r9 = true, .r10 = true, .r11 = true, .memory = true });
}

fn signalEndpoint(endpoint_id: u64) u64 {
    return asm volatile (
        \\int $0x80
        : [ret] "={rax}" (-> u64),
        : [nr] "{rax}" (syscall_signal_endpoint),
          [arg0] "{rdi}" (endpoint_id),
        : .{ .rcx = true, .rdx = true, .r8 = true, .r9 = true, .r10 = true, .r11 = true, .memory = true });
}

fn grantCap(paddr: u64, target_process_slot: u64, rights: u64) u64 {
    return asm volatile (
        \\int $0x80
        : [ret] "={rax}" (-> u64),
        : [nr] "{rax}" (syscall_grant_cap),
          [arg0] "{rdi}" (paddr),
          [arg1] "{rsi}" (target_process_slot),
          [arg2] "{rdx}" (rights),
        : .{ .rcx = true, .rdx = true, .r8 = true, .r9 = true, .r10 = true, .r11 = true, .memory = true });
}

fn installEndpoint(endpoint_id: u64, target_process_slot: u64) u64 {
    return asm volatile (
        \\int $0x80
        : [ret] "={rax}" (-> u64),
        : [nr] "{rax}" (syscall_install_endpoint),
          [arg0] "{rdi}" (@as(u64, 0)),
          [arg1] "{rsi}" (endpoint_id),
          [arg2] "{rdx}" (target_process_slot),
        : .{ .rcx = true, .rdx = true, .r8 = true, .r9 = true, .r10 = true, .r11 = true, .memory = true });
}

fn waitEvent(wait_mailbox: bool, timeout_ticks: u64) u64 {
    return asm volatile (
        \\int $0x80
        : [ret] "={rax}" (-> u64),
        : [nr] "{rax}" (syscall_wait_event),
          [arg0] "{rdi}" (@as(u64, if (wait_mailbox) 1 else 0)),
          [arg1] "{rsi}" (timeout_ticks),
        : .{ .rcx = true, .rdx = true, .r8 = true, .r9 = true, .r10 = true, .r11 = true, .memory = true });
}

fn clearPage(base_va: u64) void {
    const ptr: [*]volatile u64 = @ptrFromInt(base_va);
    var i: usize = 0;
    while (i < 512) : (i += 1) {
        ptr[i] = 0;
    }
}

fn copyBytesToVolatile(dest: [*]volatile u8, src: []const u8) void {
    var i: usize = 0;
    while (i < src.len) : (i += 1) {
        dest[i] = src[i];
    }
}

fn copyVolatileBytes(src: [*]volatile u8, dest: []u8) void {
    var i: usize = 0;
    while (i < dest.len) : (i += 1) {
        dest[i] = src[i];
    }
}

fn compilerBarrier() void {
    asm volatile ("" ::: .{ .memory = true });
}

fn parseStatus(raw: i32) ?block_protocol.Status {
    return std.meta.intToEnum(block_protocol.Status, raw) catch null;
}

fn statusToError(status: block_protocol.Status) Error {
    return switch (status) {
        .ok => unreachable,
        .invalid => error.Invalid,
        .not_found => error.NotFound,
        .no_right => error.NoRight,
        .too_big => error.TooBig,
        .io_error => error.IoError,
        .busy => error.Busy,
    };
}

File: userland/esp_server/src/block_client.c
#include "block_client.h"

enum {
    SYSCALL_GRANT_CAP = 0x08,
    SYSCALL_SIGNAL_ENDPOINT = 0x2C,
    SYSCALL_GET_PROCESS_SLOT = 0x2E,
    SYSCALL_INSTALL_ENDPOINT = 0x26,
    SYSCALL_WAIT_EVENT = 0x17,
    SYSCALL_ALLOC_MAP_PAGES_ANYWHERE = 0x5D,
    SYSCALL_CREATE_IPC_BUFFER_FROM_PAGE = 0x5E,
    SYSCALL_GRANT_IPC_BUFFER_ON_ENDPOINT = 0x5F,
    SYSCALL_SHARE_IPC_BUFFER_ON_ENDPOINT = 0x60,
    SYSCALL_OK = 0,
    SYSCALL_ERR_ENDPOINT = 9,

    IPC_BUFFER_TOKEN_TAG = 0xA000000000000000ULL,
    IPC_BUFFER_TOKEN_MASK = 0x0FFFFFFFFFFFFFFFULL,
    IPC_BUFFER_RIGHT_READ = 0x1,
    IPC_BUFFER_RIGHT_WRITE = 0x2,
    IPC_BUFFER_RIGHT_MAP = 0x4,
    IPC_BUFFER_RIGHT_GRANT = 0x8,
    IPC_BUFFER_ROLE_REQUEST = 1,
    IPC_BUFFER_ROLE_RESPONSE = 2,

    SERVICE_REGISTRY_PAGE_VA = 0x3C2C0000,
    RESPONSE_POLL_LIMIT = 512,
};

static u64 syscall0(u64 nr) {
    u64 ret;
    __asm__ volatile(
        "int $0x80"
        : "=a"(ret)
        : "a"(nr)
        : "rcx", "rdx", "r8", "r9", "r10", "r11", "memory");
    return ret;
}

static u64 syscall2(u64 nr, u64 arg0, u64 arg1) {
    u64 ret;
    __asm__ volatile(
        "int $0x80"
        : "=a"(ret)
        : "a"(nr), "D"(arg0), "S"(arg1)
        : "rcx", "rdx", "r8", "r9", "r10", "r11", "memory");
    return ret;
}

static u64 syscall3(u64 nr, u64 arg0, u64 arg1, u64 arg2) {
    u64 ret;
    __asm__ volatile(
        "int $0x80"
        : "=a"(ret)
        : "a"(nr), "D"(arg0), "S"(arg1), "d"(arg2)
        : "rcx", "r8", "r9", "r10", "r11", "memory");
    return ret;
}

static void clear_page(u64 va) {
    volatile u64 *p = (volatile u64 *)va;
    for (u64 i = 0; i < 512; i++) p[i] = 0;
}

static void copy_bytes_from_volatile(const volatile u8 *src, u8 *dst, u64 len) {
    for (u64 i = 0; i < len; i++) dst[i] = src[i];
}

static int find_block_service(u64 *endpoint_id, u64 *process_slot) {
    volatile struct service_registry_page *page = (volatile struct service_registry_page *)SERVICE_REGISTRY_PAGE_VA;
    if (page->magic != SERVICE_REGISTRY_MAGIC || page->version != SERVICE_REGISTRY_VERSION) return 0;
    u64 count = page->entry_count;
    if (count > 8) count = 8;
    for (u64 i = 0; i < count; i++) {
        volatile struct service_registry_entry *entry = &page->entries[i];
        if (entry->kind != SERVICE_KIND_BLOCK) continue;
        if (entry->endpoint_id == 0) return 0;
        *endpoint_id = entry->endpoint_id;
        *process_slot = ((entry->flags & SERVICE_FLAG_PROCESS_SLOT_COMPAT) != 0) ? entry->process_slot : 0;
        return 1;
    }
    return 0;
}

static int install_compat_endpoint_if_needed(u64 endpoint_id, u64 process_slot) {
    if (process_slot == 0) return 0;
    return syscall3(SYSCALL_INSTALL_ENDPOINT, 0, endpoint_id, process_slot) == SYSCALL_OK;
}

static int is_ipc_buffer_token(u64 token) {
    return (token & ~IPC_BUFFER_TOKEN_MASK) == IPC_BUFFER_TOKEN_TAG && (token & IPC_BUFFER_TOKEN_MASK) != 0;
}

static u64 create_ipc_buffer_from_page(u64 paddr, u64 rights_bits, u64 role) {
    return syscall3(SYSCALL_CREATE_IPC_BUFFER_FROM_PAGE, paddr, rights_bits, role);
}

static u64 grant_ipc_buffer_on_endpoint(u64 token, u64 endpoint_id, u64 rights_bits, u64 process_slot) {
    u64 ret = syscall3(SYSCALL_GRANT_IPC_BUFFER_ON_ENDPOINT, token, endpoint_id, rights_bits);
    if (is_ipc_buffer_token(ret)) return ret;
    if (ret != SYSCALL_ERR_ENDPOINT) return 0;
    if (!install_compat_endpoint_if_needed(endpoint_id, process_slot)) return 0;
    ret = syscall3(SYSCALL_GRANT_IPC_BUFFER_ON_ENDPOINT, token, endpoint_id, rights_bits);
    return is_ipc_buffer_token(ret) ? ret : 0;
}

static int share_ipc_buffer_on_endpoint(u64 token, u64 endpoint_id, u64 rights_bits, u64 process_slot) {
    u64 ret = syscall3(SYSCALL_SHARE_IPC_BUFFER_ON_ENDPOINT, token, endpoint_id, rights_bits);
    if (ret == SYSCALL_OK) return 1;
    if (ret != SYSCALL_ERR_ENDPOINT) return 0;
    if (!install_compat_endpoint_if_needed(endpoint_id, process_slot)) return 0;
    return syscall3(SYSCALL_SHARE_IPC_BUFFER_ON_ENDPOINT, token, endpoint_id, rights_bits) == SYSCALL_OK;
}

static u64 make_token_session_nonce(u64 request_token, u64 response_token, u64 endpoint_id, u64 process_slot) {
    u64 nonce = request_token ^ ((response_token << 17) | (response_token >> 47)) ^
        ((endpoint_id << 7) | (endpoint_id >> 57)) ^ process_slot ^ 0x517cc1b727220a95ULL;
    return nonce == 0 ? 1 : nonce;
}

static int wait_response(struct block_client *client, u64 seq, u16 op) {
    volatile struct block_response_header *response = (volatile struct block_response_header *)client->response_va;
    for (u64 i = 0; i < RESPONSE_POLL_LIMIT; i++) {
        if (response->response_seq == seq) {
            return response->magic == BLOCK_RESPONSE_MAGIC &&
                response->version == BLOCK_PROTOCOL_VERSION &&
                response->op == op &&
                response->status == BLOCK_STATUS_OK;
        }
        (void)syscall2(SYSCALL_WAIT_EVENT, 0, 1);
    }
    return 0;
}

static u64 begin_request(struct block_client *client, u16 op, u64 object_token, u64 block_index, u32 block_count) {
    clear_page(client->request_va);
    volatile struct block_request_header *request = (volatile struct block_request_header *)client->request_va;
    request->magic = BLOCK_REQUEST_MAGIC;
    request->version = BLOCK_PROTOCOL_VERSION;
    request->op = op;
    request->object_token = object_token;
    request->block_index = block_index;
    request->block_count = block_count;
    request->flags = 0;
    request->inline_bytes = 0;
    request->reserved0 = 0;
    request->reserved1 = 0;
    request->arg0 = 0;
    request->arg1 = 0;
    request->session_nonce = client->session_nonce;
    u64 seq = client->next_seq++;
    __asm__ volatile("" ::: "memory");
    request->request_seq = seq;
    (void)syscall2(SYSCALL_SIGNAL_ENDPOINT, client->endpoint_id, 0);
    return seq;
}

int block_client_connect(struct block_client *client) {
    u64 endpoint_id = 0;
    u64 process_slot = 0;
    if (!find_block_service(&endpoint_id, &process_slot)) return 0;

    u64 paddrs[2] = {0, 0};
    const u64 request_va = syscall3(SYSCALL_ALLOC_MAP_PAGES_ANYWHERE, 2, 1, (u64)paddrs);
    const u64 response_va = request_va + BLOCK_PAGE_BYTES;
    if (request_va < 0x1000) return 0;
    if (paddrs[0] < 0x1000 || paddrs[1] < 0x1000) return 0;
    const u64 owner_rights = IPC_BUFFER_RIGHT_READ | IPC_BUFFER_RIGHT_WRITE | IPC_BUFFER_RIGHT_MAP | IPC_BUFFER_RIGHT_GRANT;
    const u64 request_token = create_ipc_buffer_from_page(paddrs[0], owner_rights, IPC_BUFFER_ROLE_REQUEST);
    const u64 response_token = create_ipc_buffer_from_page(paddrs[1], owner_rights, IPC_BUFFER_ROLE_RESPONSE);
    if (!is_ipc_buffer_token(request_token) || !is_ipc_buffer_token(response_token)) return 0;
    const u64 remote_response_token = grant_ipc_buffer_on_endpoint(
        response_token,
        endpoint_id,
        IPC_BUFFER_RIGHT_READ | IPC_BUFFER_RIGHT_WRITE | IPC_BUFFER_RIGHT_MAP,
        process_slot
    );
    if (!is_ipc_buffer_token(remote_response_token)) return 0;

    u64 self_slot = syscall0(SYSCALL_GET_PROCESS_SLOT);
    u64 nonce = make_token_session_nonce(request_token, response_token, endpoint_id, self_slot);

    clear_page(request_va);
    clear_page(response_va);
    volatile struct block_request_header *request = (volatile struct block_request_header *)request_va;
    request->magic = BLOCK_REQUEST_MAGIC;
    request->version = BLOCK_PROTOCOL_VERSION;
    request->op = BLOCK_OP_CONNECT;
    request->request_seq = 1;
    request->arg0 = remote_response_token;
    request->arg1 = self_slot;
    request->session_nonce = nonce;

    if (!share_ipc_buffer_on_endpoint(request_token, endpoint_id, IPC_BUFFER_RIGHT_READ | IPC_BUFFER_RIGHT_MAP, process_slot)) return 0;

    client->request_va = request_va;
    client->response_va = response_va;
    client->request_paddr = paddrs[0];
    client->response_paddr = paddrs[1];
    client->request_token = request_token;
    client->response_token = response_token;
    client->endpoint_id = endpoint_id;
    client->server_process_slot = process_slot;
    client->session_nonce = nonce;
    client->next_seq = 2;

    if (!wait_response(client, 1, BLOCK_OP_CONNECT)) return 0;
    volatile struct block_response_header *response = (volatile struct block_response_header *)client->response_va;
    client->root_token = response->result_token;
    client->block_size = response->arg0;
    client->capacity_blocks = response->arg1;
    return client->root_token != 0 && client->block_size != 0 && client->capacity_blocks != 0;
}

int block_client_identify(struct block_client *client) {
    u64 seq = begin_request(client, BLOCK_OP_IDENTIFY, client->root_token, 0, 0);
    if (!wait_response(client, seq, BLOCK_OP_IDENTIFY)) return 0;
    volatile struct block_response_header *response = (volatile struct block_response_header *)client->response_va;
    client->block_size = response->arg0;
    client->capacity_blocks = response->arg1;
    return client->block_size != 0 && client->capacity_blocks != 0;
}

int block_client_read_one(struct block_client *client, u64 block_index, u8 *out, u64 out_len) {
    if (!out || out_len < client->block_size || client->block_size > BLOCK_PAGE_BYTES) return 0;
    u64 seq = begin_request(client, BLOCK_OP_READ_BLOCKS, client->root_token, block_index, 1);
    if (!wait_response(client, seq, BLOCK_OP_READ_BLOCKS)) return 0;
    volatile struct block_response_header *response = (volatile struct block_response_header *)client->response_va;
    if (response->inline_bytes != client->block_size) return 0;
    const volatile u8 *payload = (const volatile u8 *)(client->response_va + sizeof(struct block_response_header));
    copy_bytes_from_volatile(payload, out, client->block_size);
    return 1;
}
